### PR TITLE
feat(blobstream): utilizes sdk.Dec type for power difference calculation 

### LINKED ISSUE #2699: Improper Use of Floats in PowerDiff
Context: Precision and performance concerns in financial computations.
Location: [validator.go](https://github.com/celestiaorg/celestia-app/blob/main/x/qgb/types/validator.go#L115)
Issue: The method uses floats.
Remediation: Switch to using `sdk.Dec`

### LINKED ISSUE #2700: Use of Floats in SignificantPowerDiff
Context: Precision issues with floating-point numbers.
Location: [abci.go](https://github.com/celestiaorg/celestia-app/blob/main/x/qgb/abci.go#L18)
Issue: Floats are used for the `SignificantPowerDiff` constant.
Remediation: Transition to `sdk.Dec`.

### PR DESCRIPTION
**Pull request by staheri14**
Closes #2700 
Closes #2699

### CODE REVIEW

File: x/blobstream/types/validator.go
Comment (staheri14) [inline x/blobstream/types/validator.go:118]: [note to the reviewers] Alternatively, we could perform all calculations using `sdk.Dec` from the outset. However, since there are no floating-point calculations involved until later in this function, I thought it might be more efficient to continue using integer operations until that point. Nevertheless, if it is preferable to use `sdk.Dec` arithmetic throughout, I can update the initial portion of the function accordingly.

File: x/blobstream/types/validator.go
Comment (staheri14) [inline x/blobstream/types/validator.go:121]: The validator powers are of type `uint64`, and the current conversion could potentially lead to overflow. Upon reviewing the [`validateBasics`](https://github.com/celestiaorg/celestia-app/blob/6683a049b95729c2d055c8e08e3c81d9b7bb7914/x/blobstream/types/validator.go#L57) function, it appears that there is no range check performed on the power value. It would be prudent to confirm whether there might be any validators with powers falling within the range of [`math.MaxInt64`, `math.MaxUint64`]. Such power values could indeed cause an overflow issue in the code line mentioned above. (For example the power value of `math.MaxUint64` becomes `-1` when converted to `int64`)

**Comment by staheri14** [comment]
It appears that this PR also resolves issue #2699. I couldn't locate any `float` calculations in validator.go except for those being handled by this PR. @SweeXordious, could you please confirm whether this is the case? then I'll update the description of PR accordingly.

File: x/blobstream/types/validator.go
Comment (SweeXordious) [reply x/blobstream/types/validator.go:121]: good catch :+1: But I think if the state machine got to a state where this overflows, then something is seriously messed up with the staking module. In fact, values like those will not be accepted by the staking module, then we normalize those values before this step. I think that we can use a validation at this level as a last line of defense, but reaching it is super unlikely if I understand corerctly

**Comment by SweeXordious** [comment]
> It appears that this PR also resolves issue https://github.com/celestiaorg/celestia-app/issues/2699. I couldn't locate any float calculations in validator.go except for those being handled by this PR. @SweeXordious, could you please confirm whether this is the case? then I'll update the description of PR accordingly.

Apparently yes, this is also handling that issue, please do update. Thanks a lot for helping with this :pray:

File: x/blobstream/types/validator.go
Comment (SweeXordious) [reply x/blobstream/types/validator.go:118]: I think switching to `sdk.Dec` would be also future-proof in case we wanted to make any changes not to give to future devs the ability to choose. So, I would go with switching everything to `sdk.Dec`

**Review by SweeXordious — COMMENTED** [review]
Awesome stuff :+1: 

I will approve after we switch all arithmetic operations to using `sdk.Dec`.

File: x/blobstream/types/validator.go
Comment (staheri14) [reply x/blobstream/types/validator.go:121]: > values like those will not be accepted by the staking module

So, does it mean that there will never be a case in which a validator's power reside in this range [math.MaxInt64, math.MaxUint64]? and it is guaranteed by the staking module?

Comment (SweeXordious) [reply x/blobstream/types/validator.go:121]: yes, the validator power is using `int64`, so no overflow would happen. However, I found something interesting:
https://github.com/celestiaorg/cosmos-sdk/blob/main/x/staking/types/validator.go#L369-L372

```
// PotentialConsensusPower returns the potential consensus-engine power.
func (v Validator) PotentialConsensusPower(r math.Int) int64 {
	return sdk.TokensToConsensusPower(v.Tokens, r)
}
```

The power is being reduced by a factor of:
```
// DefaultPowerReduction is the default amount of staking tokens required for 1 unit of consensus-engine power
var DefaultPowerReduction = NewIntFromUint64(1000000)
```

However, the tokens amount is represented by an int256... So, there might be overflow there when going from tokens to consensus-engine power unless there is the assumption that the staked tokens wouldn't be that much.

cc @evan-forbes any idea?

Comment (evan-forbes) [reply x/blobstream/types/validator.go:121]: sorry mind pointing out where's the int256 to save me a sec :sweat_smile: 

this is good to keep in the back of our heads, especially for networks such as arabica where the supply and voting power is getting rather insane. for normal networks tho the assumption of staying within the max and min values appears safe afaiu (please correct me if I'm wrong I didn't deep dive the code yet)

**Comment by evan-forbes** [comment]
> I think switching to sdk.Dec would be also future-proof in case we wanted to make any changes not to give to future devs the ability to choose. So, I would go with switching everything to sdk.Dec

imo I think we should do this, but in a different issue/bump. We still need to test this on IRL networks to make sure that we're doing this in a non consensus breaking way or reevaluate and the simpler we can make that process the better

File: x/blobstream/types/validator.go
Comment (evan-forbes) [inline x/blobstream/types/validator.go:140]: could we leave delta a int64 the entire time?

**Review by evan-forbes — DISMISSED** [review]
one quick question to remove another instance of floats right above these, but still not blocking on it

**Comment by evan-forbes** [comment]
before merging the backported version of this, we should sync from scratch and ideally perform other tests to ensure that this is not consensus breaking. If it is, then we can reevaluate our options from there

File: x/blobstream/types/validator.go
Comment (SweeXordious) [reply x/blobstream/types/validator.go:121]: In the [`validator`](https://github.com/celestiaorg/cosmos-sdk/blob/4c897c6e15c7c0b956ca9fa25d0625ff7a39553a/x/staking/types/staking.pb.go#L314-L339) struct, the delegated tokens are represented using [`Tokens`](https://github.com/celestiaorg/cosmos-sdk/blob/4c897c6e15c7c0b956ca9fa25d0625ff7a39553a/x/staking/types/staking.pb.go#L324C26-L324C26) which are [`Int`](https://github.com/celestiaorg/cosmos-sdk/blob/4c897c6e15c7c0b956ca9fa25d0625ff7a39553a/types/math.go#L14)s which are [`sdkmath.Int`](https://github.com/celestiaorg/cosmos-sdk/blob/4c897c6e15c7c0b956ca9fa25d0625ff7a39553a/math/int.go#L72-L77). 

Then, these are used when calculating the validator delegation and power in [here](https://github.com/celestiaorg/cosmos-sdk/blob/4c897c6e15c7c0b956ca9fa25d0625ff7a39553a/x/staking/keeper/val_state_change.go#L169) after getting the `consensus-engine power` which would [panic](https://github.com/celestiaorg/cosmos-sdk/blob/4c897c6e15c7c0b956ca9fa25d0625ff7a39553a/math/int.go#L156-L161) when converting from `sdkmath.Int` to `int64` if the amount is out of bounds.

File: x/blobstream/types/validator.go
Comment (staheri14) [reply x/blobstream/types/validator.go:140]: @evan-forbes Is the concern related to the `float64` conversion in the earlier lines? If so, then the short answer is yes, we can keep `delta` as `int64` the entire time.

The reason `delta` has been defined as `float64` is because `math.Abs()` accepts and outputs `float64` values. 

If the intention is to keep `delta` as an `int64`, there are two alternative approaches to achieve that:

1. Maintain `delta` as an `int64`, but continue using `math.Abs()` by converting individual powers to `float64`:
```go
var delta int64
for _, v := range powers {
    // NOTE: We are interested in the absolute value of the changes
    delta += int64(math.Abs(float64(v)))
}
```

2. Completely avoid the `float64` conversion of the powers by not using `math.Abs()`. However, please be aware that this approach may have undefined behavior in handling corner cases like when `v` is equal to ±Inf or NaN (which `math.Abs()` can handle: [Link](https://pkg.go.dev/math#Abs)):
```go
var delta int64
for _, v := range powers {
    // NOTE: We are interested in the absolute value of the changes
    if v < 0 {
        v = -v
    }
    delta += int64(v)
}
```
I lean towards option 2, only if the corner cases I mentioned are not conceivable. 
cc: @SweeXordious

**Comment by staheri14** [comment]
> > I think switching to sdk.Dec would be also future-proof in case we wanted to make any changes not to give to future devs the ability to choose. So, I would go with switching everything to sdk.Dec

> imo I think we should do this, but in a different issue/bump. We still need to test this on IRL networks to make sure that we're doing this in a non consensus breaking way or reevaluate and the simpler we can make that process the better


Considering this comment of @evan-forbes, do we still want to refactor the initial section of the `PowerDiff` function to utilize `sdk.Dec` @SweeXordious in this PR? In my opinion, since the initial section of this function exclusively involves integer operations, converting it to utilize `sdk.Dec` arithmetic wouldn't enhance computational precision; instead, it would primarily improve code readability and clarity. 
PS: I have already started updating the first portion of the `PowerDiff` and it is in progress; there is one intricacy with the `Sub` method of sdk.Dec type, which converts zero values to `nil` which then causes the rest of computations to panic (and I am debugging and trying to find a fix). I can open a follow-up PR  to refactor the first portion of `PowerDiff` function, if there is an agreement on whether refactoring is needed or not. 

>  I think switching to sdk.Dec would be also future-proof in case we wanted to make any changes not to give to future devs the ability to choose.

I'm unsure how this change would effectively enforce any constraints on future developers when the function signature remains unchanged, and only its internals  are to be modified to utilize `sdk.Dec`. Could you please provide more insight into your perspective?

File: x/blobstream/types/validator.go
Comment (staheri14) [reply x/blobstream/types/validator.go:118]: Thanks for your comment! 
I'm unsure how this change would effectively enforce any constraints on future developers when the function signature remains unchanged, and only its internals are to be modified to utilize `sdk.Dec`. Could you please provide more insight into your perspective?

File: x/blobstream/types/validator.go
Comment (evan-forbes) [reply x/blobstream/types/validator.go:140]: yeah option 2 is what I think would be preferable

**Comment by evan-forbes** [comment]
> In my opinion, since the initial section of this function exclusively involves integer operations, converting it to utilize sdk.Dec arithmetic wouldn't enhance computational precision; instead, it would primarily improve code readability and clarity.

my understanding was that this would potentially stop an overflow with large stake? correct me if I'm wrong @SweeXordious

reguardless I think we should merge this or some version of it sooner rather than later as ideally this would be included in the version that is used for mainnet

**Comment by SweeXordious** [comment]
> my understanding was that this would potentially stop an overflow with large stake? correct me if I'm wrong @SweeXordious

Yes, that's the idea. But after investigating the cosmos-sdk implementation, this would fail in the staking module prior to reaching this level, since they're not handling the out of bounds case. source: https://github.com/celestiaorg/celestia-app/pull/2719#discussion_r1365261371

However, I agree that we should ship some version of this to be included in the upcoming release so that this change is also audited.

> Completely avoid the float64 conversion of the powers by not using math.Abs(). However, please be aware that this approach may have undefined behavior in handling corner cases like when v is equal to ±Inf or NaN (which math.Abs() can handle: [Link](https://pkg.go.dev/math#Abs)):

I also agree with going more with option 2. And we could even add explicit checks to see if `v` is out of bounds or `NaN` if that's possible.

Thanks a lot @staheri14 for helping with this stuff :pray:

**Comment by staheri14** [comment]
I am working on option 2, will soon commit my changes.

File: x/blobstream/types/validator.go
Comment (staheri14) [inline x/blobstream/types/validator.go:158]: [Note to reviwers @evan-forbes @SweeXordious ] This is a tricky and problematic corner case, without this line, the negative of `-9223372036854775808` (math,MinInt64) equals itself but not `9223372036854775808`. The reason is explained in the function description.


Upon further thoughts, another issue came to my mind: if the sum of the absolute power diffs exceeds the maximum value representable by `int64` then there will be overflow issues (do we expect this happen?). This is while using `float64` it is less likely since the largest value representable by `float64` is much larger than the max value that an `int64` can accomodate. So, we may want to reconsider the original version in which delta was `float64` and we have been using math.Abs() instead. Please let me know your thoughts.

Comment (staheri14) [reply x/blobstream/types/validator.go:158]: Please note that float64 calculations also introduce their own errors, as illustrated below:

```go
v1 := math.MinInt64 // -9223372036854775808
v2 := float64(math.MinInt64) // -9223372036854776000
```

When examining the values of `v1` and `v2`, it becomes evident that the `float64` representation of `v1` does not precisely match its `int64` value.

Comment (evan-forbes) [reply x/blobstream/types/validator.go:158]: I don't think we have anything to worry about as long as we use normal values (most testnets have 1*10^15 total utia and 1*10^9 voting power), this is just being extra careful. 

I'm definitely okay with shipping and testing smaller changes, and was incorrect to assume that staying with int64 was a straightforward switch, will defer to the author on that

Comment (evan-forbes) [reply x/blobstream/types/validator.go:158]: I'm tempted to just use big.Int for everything

Comment (SweeXordious) [reply x/blobstream/types/validator.go:158]: Hum, yes `big.Int` or we just use `sdk.Dec` wrapper for everything, even the map of powers created in `powers := map[string]int64{}`. That would help us avoid all of these conversions but not sure about the consequences of it in functions calling the `PowerDiff`.

**Review by evan-forbes — APPROVED** [review]

File: x/blobstream/types/validator.go
Comment (staheri14) [reply x/blobstream/types/validator.go:158]: In the setting you mentioned i.e.,
> (most testnets have 1x10^15 total utia and 1x10^9 voting power)

it should be fine to use `int64` to hold the sum of power diffs, but if in Arabica it can get larger and the upper bound is not known yet, then I'd vote for keeping delta as `float64` just to avoid potential overflows.

@SweeXordious Once you mentioned powers are normalized? to what range they are normalized to? What is the max power value after normalization? Basically, what is the max power value that `PowerDiff` function can receive as its input? Is the sum of powers capped?

> I'm tempted to just use big.Int for everything

This or `sdk.Dec` which underneath is a `big.Int`.

Comment (SweeXordious) [reply x/blobstream/types/validator.go:158]: > @SweeXordious Once you mentioned powers are normalized, to what range they are normalized to? What is the max power value after normalization? Basically, what is the max power value that PowerDiff function can receive as its input? Is the sum of powers capped?

They're normalized to `2^32`:
https://github.com/celestiaorg/celestia-app/blob/fc348bf395e0c56787aa0fe587275adf2f04c40b/x/blobstream/keeper/keeper_valset.go#L136-L161

Comment (staheri14) [reply x/blobstream/types/validator.go]: Thank you for providing the context. While I believe that using `int64` with this normalization would have also been functional, I have since successfully refactored the `PowerDiff` function to utilize `sdk.Dec` from the outset (managed to resolve the bug I mentioned earlier). Regarding the code review, everything seems to be in order, except for the potential overflow issue when casting from `uint64` to `int64`; we have partially discussed it, however, it is out of scope of this PR and merits its own separate PR.

File: x/blobstream/types/validator.go
Comment (SweeXordious) [inline x/blobstream/types/validator.go:121]: we can avoid this unsafe conversion using `sdk.NewDecFromBigInt(new(big.Int).SetUint64(bv.Power))`

Comment (staheri14) [reply x/blobstream/types/validator.go]: Thanks, done https://github.com/celestiaorg/celestia-app/pull/2719/commits/efae96232a7187b32179e37e298fddb4348b15b1.

File: .github/workflows/ci-release.yml
Comment (SweeXordious) [inline .github/workflows/ci-release.yml:53]:
```suggestion
    needs: [lint, markdown-linter, test]
```
probably?

File: .goreleaser.yaml
Comment (SweeXordious) [inline .goreleaser.yaml:20]:
```suggestion
```
I guess some other PR got confused into this

**Review by SweeXordious — COMMENTED** [review]
awesome stuff :+1: :+1: 

just apply the comments and will approve

File: .github/workflows/ci-release.yml
Comment (staheri14) [reply .github/workflows/ci-release.yml:53]: I did not make this change!

Comment (SweeXordious) [reply .github/workflows/ci-release.yml:53]: This is funny :D :D

Probably wrong PR was targetted: https://github.com/celestiaorg/celestia-app/pull/2719/commits/21f244af1be156ba9891c21ff443759121b0f5ae

cc  @rootulp

File: x/blobstream/types/validator.go
Comment (staheri14) [reply x/blobstream/types/validator.go]: Actually this commit is the correct one https://github.com/celestiaorg/celestia-app/pull/2719/commits/bddfcfa3823718def693ee3293bb57183d0512d7

File: .github/workflows/ci-release.yml
Comment (staheri14) [reply .github/workflows/ci-release.yml]: I merged a wrong branch into my branch, fixed it now!

File: x/blobstream/types/validator.go
Comment (SweeXordious) [inline x/blobstream/types/validator.go:128]:
```suggestion
		bvPower := sdk.NewDecFromBigInt(new(big.Int).SetUint64(bv.Power))
```

File: .goreleaser.yaml
Comment (staheri14) [reply .goreleaser.yaml]: Indeed, I merged a wrong branch, though I fixed it!

File: x/blobstream/types/validator.go
Comment (staheri14) [reply x/blobstream/types/validator.go]: Done now https://github.com/celestiaorg/celestia-app/pull/2719/commits/d3c7ae0a54fd68413d3ad8072aa8dbd371df8f97

**Review by SweeXordious — APPROVED** [review]
:shipit:

**Review by evan-forbes — APPROVED** [review]

**Pull request merged**
